Rename select* functions to partialsort* and various related fixes

nalimilan opened the pull request:
The rename (first commit) fixes #22791. The second commit moves docstrings out of helpdb. The third commit fixes what appears as a documentation bug.

The fourth one changes `selectperm!`/`partialsortperm!` to return a view rather than a copy to avoid an allocation (which is really silly if the number of elements to sort is relatively large). Another possibility would be to return the input array just like `sortperm!` does, since the caller has enough information to extract the relevant elements. But that doesn't sound useful, and the cost of creating a view is quite small.

I've kept the commits separate to make it easier to distinguish the actual changes from mere renaming and moving operations. They don't need to be squashed when merging since they should all pass the tests.

Closes #22791.

nalimilan committed:
Rename select* functions to partialsort*

The new name is more explicit, more consistent with sort and with one of the
most commonly used names for this operation (from the C++ stdlib). It also
does not conflict with other meanings e.g. for POSIX sockets and SQL.

nalimilan committed:
Make partialsort!() and partialsortperm!() return a view rather than a copy

Returning a copy (partially) defeats the purpose of these functions, which is
to avoid allocations.

tkelman commented on base/sort.jl:
```diff
@@ -609,18 +668,35 @@ julia> v
+    partialsortperm!(ix, v, k, [alg=<algorithm>,] [by=<transform>,] [lt=<comparison>,] [rev=false,] [initialized=false])
+
+Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix`. If `initialized` is `false`
+(the default), ix is initialized to contain the values `1:length(ix)`.
```
code highlight `ix`

nalimilan replied:
Good catch. Fixed now.

kmsquire commented:
I'm wondering about the status of using named tuples as part of dispatch (#22194)?

`partialsort` is reasonable as a name, but `partialsortperm` is starting to get a little unwieldy, in my opinion.  I think this functionality would be a bit nicer as a keyword argument to `sort` / `sortperm` (and friends):

`sort(x, partial=1:5)`

or even

`sort(x, select=1:5)`

Thoughts?

nalimilan commented:
Actually, if we want to merge `select` into `sort`, we could just ask users to pass `PartialQuickSort(k)` as `alg`. A `partial` keyword argument would also work. A possible drawback is that it would only work with `alg=QuickSort` (i.e. replacing it with `PartialQuickSort` under the hood), and I'm not sure whether partial sorting variants exist for other sorting algorithms.

kmsquire commented:
I like the idea of a partial keyword, even though it would only work with quicksort.

nalimilan commented:
Thinking about it again, a `partial` keyword argument would have the additional drawback that the `alg`orithm type would be computed at runtime, so `sort`/`sortperm` would always have to go via dynamic dispatch. That would incur a performance penalty even when not passing `partial=k` explicitly. This could be problematic when sorting small arrays.

So I'd rather go with separate `partialsort` and `partialsortperm` functions, or remove the functions and require people to pass `PartialQuickSort(k)` via the `alg` argument.

Thoughts?

JeffBezanson commented on base/sort.jl:
```diff
@@ -82,16 +82,80 @@ issorted(itr;
+same thing as `partialsort!` but leaving `v` unmodified.
+"""
+partialsort(v::AbstractVector, k::Union{Int,OrdinalRange}; kws...) =
+    copy(partialsort!(copymutable(v), k; kws...))
```
I would remove the call to `copy`. In the `k isa Integer` case, it means the sorted objects need to have `copy` methods, which should not be required to sort them.

nalimilan replied:
Good catch. OTOH it feels a bit weird to return a view of something which the user never had access to. But it doesn't really hurt either, and that's the most efficient solution (the user can still can `copy` manually if needed). I've updated the PR to drop `copy`.

kmsquire commented:
Since `alg` is a keyword argument, isn't dispatch done at runtime anyway?

nalimilan commented:
> Since alg is a keyword argument, isn't dispatch done at runtime anyway?

Yes, currently that's how it works, but I think there are plans to change this. @JeffBezanson can confirm or infirm that.

JeffBezanson commented:
I think sorting and partial sorting are different enough to warrant separate functions.

kmsquire commented:
`sort` and `partialsort` are fine--it's `partialsortperm` that I was (originally) objecting to when I suggested merging the functions.

kmsquire commented:
But anyway, I don't have any further ideas right now, so if others are fine with changing `selectperm` to `partialsortperm`, go for it.

tkelman commented on base/deprecated.jl:
```diff
@@ -1632,6 +1632,12 @@ function SymTridiagonal(dv::AbstractVector{T}, ev::AbstractVector{S}) where {T,S
 end
 
+# issue #22791
+@deprecate_binding select partialsort
```
don't you only need `@deprecate_binding` for types/modules? isn't `@deprecate` good enough for functions?

nalimilan replied:
Yeah, I'm not sure what's the difference. Since the function itself is going to be removed (not only some methods), I figured deprecating the binding makes sense. Are there any unintended consequences?

fredrikekre replied:
You get different stack traces, and to some extent different information:
```julia
julia> foo() = 1;

julia> Base.@deprecate_binding bar foo

julia> bar()
WARNING: Main.bar is deprecated, use Main.foo instead.
  likely near no file:0
1
```

vs

```julia
julia> foo() = 1;

julia> Base.@deprecate bar foo
:bar

julia> bar()
WARNING: bar is deprecated, use foo instead.
Stacktrace:
 [1] depwarn at ./deprecated.jl:70 [inlined]
 [2] bar() at ./deprecated.jl:31
 [3] eval(::Module, ::Expr) at ./repl/REPL.jl:3
 [4] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./repl/REPL.jl:69
 [5] macro expansion at ./repl/REPL.jl:100 [inlined]
 [6] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
while loading no file, in expression starting on line 0
1
```

nalimilan replied:
OK, I don't know the reason for this difference but that's enough to simply use `@deprecate`. PR updated.

tkelman replied:
deprecate_binding works at the C level and is able to detect more uses like in dispatch constraints of other methods (ref #11200), deprecate only detects calls, but that should be good enough for methods

StefanKarpinski commented:
`partialsortperm` is a pretty ugly name, but it's a very specific function, so I'm fine with it.

Sacha0 commented on base/sort.jl:
```diff
@@ -82,16 +82,80 @@ issorted(itr;
     sort!(v, first(inds), last(inds), PartialQuickSort(k), o)
-    v[k]
+
+    if k isa Integer
```
Perhaps using dispatch for this logic would be better than branching?

nalimilan replied:
Let's go with the current version then. I've rebased again, let's not wait for too long as there were already substantial conflicts.

nalimilan replied:
Actually, `@views` seems to do the trick. See https://github.com/JuliaLang/julia/pull/23760.

Sacha0 commented on base/sort.jl:
```diff
@@ -629,7 +709,12 @@ function selectperm!(ix::AbstractVector{<:Integer}, v::AbstractVector,
     sort!(ix, PartialQuickSort(k), Perm(ord(lt, by, rev, order), v))
-    return ix[k]
+
+    if k isa Integer
```
Likewise here, perhaps using dispatch for this purpose would be better than branching?

nalimilan commented on base/sort.jl:
```diff
@@ -82,16 +82,80 @@ issorted(itr;
     sort!(v, first(inds), last(inds), PartialQuickSort(k), o)
-    v[k]
+
+    if k isa Integer
```
In general I'd agree with you, but in the present case that would force duplicating the two lines above, that's why I chose the `if` approach.

BTW, there seems to be a lack in the API: with `getindex`, you can either get an array or a scalar, but with `view`/`@view` you always get a subarray. Shouldn't there be a way to get the same behavior as `getindex`? Then one wouldn't have to check types manually like this (which could be hard or impossible in more complex cases where the index can be of any type.)

Sacha0 replied:
Would not
```julia
function partialsort!(v::AbstractVector, k::Union{Int,OrdinalRange}, o::Ordering)
    inds = indices(v, 1)
    sort!(v, first(inds), last(inds), PartialQuickSort(k), o)
    return _dispatchftw(v, k)
end
_dispatchftw(v, k::Integer) = v[k]
_dispatchftw(v, k) = view(v, k)
```
do the trick? Could even reuse the helper functions for the same purpose below :).

nalimilan replied:
I guess that would work, but I don't find it clearer than the existing version, do you? To understand what the code does you need to look at the definitions of methods for `_dispatchftw`, which does not have a generic meaning on its own.

Sacha0 replied:
> I guess that would work, but I don't find it clearer than the existing version, do you?

Noting that my subjective evaluation is practically meaningless, yes, I prefer this style :).

Perhaps a more meaningful question would be whether one or the approach is friendlier to the compiler?

nalimilan replied:
Style debates are hard to decide! FWIW, I don't think it makes any difference for the compiler.

Sacha0 replied:
Out of curiosity I checked the code each implementation generates (less the irrelevant `indices` and `sort!` calls); the generated code is essentially identical :).

nalimilan replied:
Could a third person cast a vote here to decide which form is best? Let's not hold this PR off just because we can't make a decision. :-)

KristofferC replied:
I think inlining is clearer because otherwise, you have to go look for `_dispatchftw` and you might not realize it is defined just below. But I am generally of the opinion that inlining stuff is to prefer over breaking out things in function, when those functions are only used in one place.

fredrikekre replied:
I don't really care either, if it does not matter for technical reasons. But on the other hand we have: https://docs.julialang.org/en/stable/manual/performance-tips/#Break-functions-into-multiple-definitions-1 :trollface:

KristofferC replied:
There is a difference. There you break out the whole function, not just creating a tiny shim for a specific type.

fredrikekre replied:
There is always the third option, with the same lines of code:
```julia
function partialsort!(v::AbstractVector, k::Int, o::Ordering)
    inds = indices(v, 1)
    sort!(v, first(inds), last(inds), PartialQuickSort(k), o)
    return v[k]
end
function partialsort!(v::AbstractVector, k::OrdinalRange, o::Ordering)
    inds = indices(v, 1)
    sort!(v, first(inds), last(inds), PartialQuickSort(k), o)
    return view(v, k)
end
```

KristofferC replied:
The harsh reality is, not all lines of code are equal.

fredrikekre replied:
Yea, I'm voting for status quo or perhaps change it to using the ternary operator.

nalimilan committed:
Fix docstrings for partialsortperm() and partialsortperm!()

partialsort(x, k::Integer), partialsortperm(x, k::Integer) and partialsortperm!(x, k::Integer)
all return a single value. Also add missing backticks.

nalimilan commented:
I'll merge tomorrow unless somebody objects.
